### PR TITLE
Move `onCommitOrAbort` to `Transaction`

### PR DESCRIPTION
**Pull request by fsamuel-bs**
## General
**After this PR**:
Move `onCommitOrAbort()` to `Transaction` and delete `CallbackAwareTransaction`.

This is needed for a further refactor I'll do in SnapshotTransactionManager as a flup to https://github.com/palantir/atlasdb/pull/7407. Also, might be useful outside of Atlas as a primitive (e.g. clients might want to register cleanup tasks as part of transaction committing or aborting).

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@raiju

### CODE REVIEW

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java [hunk 1]
```diff
@@ -450,6 +450,27 @@ enum TransactionType {
...
+     * commit successfully.
+     * <p>
+     * Callbacks are not atomic nor transactional - it's possible for a transaction to succeed and its callbacks to not
+     * be triggered (e.g. the Java process was interrupted before).
+     * Callbacks run even if a prior callback threw an exception.
+     * Callbacks are run in the reverse order they were added - i.e. the last added callback will be run first.
+     * Different from {@link #onSuccess(Runnable)}, {@link #onCommitOrAbort(Runnable)} callbacks are run if the
+     * transaction succeeded or failed.
+     */
+    void onCommitOrAbort(Runnable runnable);
```
Comment (fsamuel-bs) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java:472]: Just moved from `CallbackAwareTransaction`.

**Review by ergo14 — COMMENTED** [review]
lgtm

**Review by jeremyk-91 — APPROVED** [review]
can we check making sure this doesn't break large internal product / if it does, fix them?

assuming we do that, 👍

**Comment by fsamuel-bs** [comment]
From a quick glance it won't.

**Pull request merged**